Source/DataLayer/EfClasses/Book.cs

```diff
@@ -144,47 +131,70 @@ private void replaceContributors(IEnumerable<Contributor> newContributors, Role
+                getEntry(context).Collection(s => s.ContributorsLink).Load();
+
+            var isIdentical
+                = ContributorsLink
+                .ByRole(role)
+                .Select(c => c.Contributor)
+                .SequenceEqual(newContributors);
 
-			var roleContributions = getContributions(role);
-            var isIdentical = roleContributions.Select(c => c.Contributor).SequenceEqual(newContributors);
             if (isIdentical)
                 return;
 
-            _contributorsLink.RemoveWhere(bc => bc.Role == role);
+            ContributorsLink.RemoveWhere(bc => bc.Role == role);
             addNewContributors(newContributors, role);
         }
 
-		private void addNewContributors(IEnumerable<Contributor> newContributors, Role role)
+        private void addNewContributors(IEnumerable<Contributor> newContributors, Role role)
         {
             byte order = 0;
             var newContributionsEnum = newContributors.Select(c => new BookContributor(this, c, role, order++));
             var newContributions = new HashSet<BookContributor>(newContributionsEnum);
-            _contributorsLink.UnionWith(newContributions);
+            ContributorsLink.UnionWith(newContributions);
         }
 
-        private List<BookContributor> getContributions(Role role)
-            => ContributorsLink
-                .Where(a => a.Role == role)
-                .OrderBy(a => a.Order)
-                .ToList();
         #endregion
 
-		private Microsoft.EntityFrameworkCore.ChangeTracking.EntityEntry<Book> getEntry(DbContext context)
-		{
-			ArgumentValidator.EnsureNotNull(context, nameof(context));
+        private Microsoft.EntityFrameworkCore.ChangeTracking.EntityEntry<Book> getEntry(DbContext context)
+        {
+            ArgumentValidator.EnsureNotNull(context, nameof(context));
 
-			var entry = context.Entry(this);
+            var entry = context.Entry(this);
 
-			if (!entry.IsKeySet)
-				throw new InvalidOperationException("Could not load a valid Book from database");
+            if (!entry.IsKeySet)
+                throw new InvalidOperationException("Could not load a valid Book from database");
 
-			return entry;
+            return entry;
+        }
+
+        #region categories
+        internal HashSet<BookCategory> CategoriesLink { get; private set; }
+
+        private ReadOnlyCollection<BookCategory> _categoriesReadOnly;
+        public ReadOnlyCollection<BookCategory> Categories
+        {
+            get
+            {
+                if (_categoriesReadOnly?.SequenceEqual(CategoriesLink) is not true)
+                    _categoriesReadOnly = CategoriesLink.ToList().AsReadOnly();
+                return _categoriesReadOnly;
+            }
+        }
+		public void SetCategoryLadders(IEnumerable<CategoryLadder> ladders)
+		{
+			ArgumentValidator.EnsureNotNull(ladders, nameof(ladders));
+
+			//Replace all existing category ladders.
+            //Some books make have duplocate ladders
```

Comment:
This little tidbit right here cost hours.

Reply:
Oof, that sounds like a fun edge case to debug